### PR TITLE
LPS-187428

### PR DESCRIPTION
**Pull request by natocesarrego**
Hi,

During the work on a few upgrades, the Hotel Team realized that there are some opportunities of improvement regarding the upgrade of FreeMarker templates. Please consider as FreeMarker templates the following resources:

- Product Menu > Content & Data > Web Content > Templates
- Product Menu > Design > Fragments
- Product Menu > Design > Templates > Widget Templates
- Product Menu > Design > Templates > Information Templates
- Product Menu > Design > Page Templates > Master
- Product Menu > Design > Page Templates > Page Templates
- Product Menu > Design > Page Templates > Display Page Templates

Basically, this pull creates a new upgrade process which visits all FreeMarker templates, looks for some patterns to be removed, replaces these patterns and injects default variables provided by Liferay Portal.

@liferay-echo, do you mind to review this pull too?

Since this pull touches on the database table `FragmentEntry` then I think that makes sense you, @liferay-echo, take a look on these changes. 

Issue link: https://liferay.atlassian.net/browse/LPS-187428

Thanks

cc @alberttgomes @velosofelipe64

### CODE REVIEW

**Comment by natocesarrego** [comment]
@liferay-forms, SF reported test failures aren't related

**Comment by marcelabc** [comment]
Just started reviewing :)

**Review by marcelabc — CHANGES_REQUESTED** [review]
Hi 
I finished my review. Let me know if you have questions :smile:

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/BaseTemplateUpgradeProcess.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+						Matcher isAssignEmptyMatcher =
+							_isAssignEmptyDDMTemplatePattern.matcher(script);
+
+						if (isAssignEmptyMatcher.find()) {
+							script = isAssignEmptyMatcher.replaceAll(
+								getTemplatePatternReplacement());
+						}
+					}
+
+					long templateId = resultSet.getLong(1);
```
[the remark this answers is not in the document]
Comment (natocesarrego) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/BaseTemplateUpgradeProcess.java:93]: Sure, @marcelabc!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/BaseTemplateUpgradeProcess.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+			PreparedStatement updatePreparedStatement =
+				AutoBatchPreparedStatementUtil.concurrentAutoBatch(
+					connection,
+					"update DDMTemplate set script = ? where templateId = ?")) {
+
+			try (ResultSet resultSet = selectPreparedStatement.executeQuery()) {
+				while (resultSet.next()) {
+					String script = resultSet.getString(2);
+
+					Matcher templateMatcher = getTemplatePattern().matcher(
```
[the remark this answers is not in the document]
Comment (natocesarrego) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/BaseTemplateUpgradeProcess.java:70]: Sure!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/BaseTemplateUpgradeProcess.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+						Matcher isAssignEmptyMatcher =
+							_isAssignEmptyFragmentEntryPattern.matcher(html);
+
+						if (isAssignEmptyMatcher.find()) {
+							html = isAssignEmptyMatcher.replaceAll(
+								getTemplatePatternReplacement());
+						}
+					}
+
+					long fragmentEntryId = resultSet.getLong(1);
```
[the remark this answers is not in the document]
Comment (natocesarrego) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/BaseTemplateUpgradeProcess.java:143]: Will do, @marcelabc!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v5_3_2/TemplateBrowserSnifferUpgradeProcess.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.liferay.dynamic.data.mapping.internal.upgrade.BaseTemplateUpgradeProcess;
+import com.liferay.petra.string.StringPool;
+
+import java.util.regex.Pattern;
+
+/**
+ * @author Albert Gomes Cabral
+ */
+public class TemplateBrowserSnifferUpgradeProcess
```
[the remark this answers is not in the document]
Comment (natocesarrego) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v5_3_2/TemplateBrowserSnifferUpgradeProcess.java:25]: We do have, @marcelabc!

Basically, since the most recent portal versions, a few variables started to be added by default in the context of the FreeMarker templates, please see all these variables in the method [prepare](https://github.com/liferay/liferay-portal/blob/f63af3907491f652f2c1eed4216b63825f26b1fc/modules/apps/portal-template/portal-template-engine-api/src/main/java/com/liferay/portal/template/engine/TemplateContextHelper.java#L202) of the class TemplateContextHelper. This pull aims to create a foundation to make easier the adding of the new cases like the variable browser sniffer.

Does it make sense to you?

[the remark this answers is not in the document]
Comment (marcelabc) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v5_3_2/TemplateBrowserSnifferUpgradeProcess.java:25]: Thank you for the explanation! Yes, it makes sense :+1:

**Comment by natocesarrego** [comment]
@marcelabc, this pull request is updated considering your valuable comments. :slightly_smiling_face: Thanks for sharing them! :smile:

**Comment by natocesarrego** [comment]
Do you have any update on this pull, @marcelabc?

Thanks

**Review by marcelabc — APPROVED** [review]
LGTM 🎉
Sorry for the delay in reviewing this PR.

**Comment by natocesarrego** [comment]
Thanks for you update, @marcelabc! :smile: 

I'll forward the pull to Brian

**Comment by natocesarrego** [comment]
ci:forward

**Comment by carolmariaabb** [comment]
ci:forward

**Comment by carolmariaabb** [comment]
ci:forward:force
